app/views/layouts/base.html.erb

```diff
@@ -40,6 +40,28 @@ See COPYRIGHT and LICENSE files for more details.
   <meta name="current_menu_item" content="<%= current_menu_item %>"/>
   <%# Disable prefetching for now %>
   <meta name="turbo-prefetch" content="false">
+  <%# Apply system theme immediately to prevent flickering %>
+  <% if User.current.pref.sync_with_os_theme? %>
+    <%= nonced_javascript_tag do %>
+      (function waitForBody() {
+      if (!document.body) {
+      return requestAnimationFrame(waitForBody);
+      }
+
+      const prefersDark = window.matchMedia('(prefers-color-scheme: dark)').matches;
+      const prefersHighContrast = window.matchMedia('(prefers-contrast: more)').matches;
+      const colorMode = prefersDark ? 'dark' : 'light';
+      const increaseContrast = prefersHighContrast;
+
+      const body = document.body;
+      const otherColorMode = colorMode === 'light' ? 'dark' : 'light';
+
+      body.setAttribute('data-color-mode', colorMode);
+      body.setAttribute(`data-${colorMode}-theme`, increaseContrast ? `${colorMode}_high_contrast` : colorMode);
+      body.removeAttribute(`data-${otherColorMode}-theme`);
+      })();
+    <% end %>
+  <% end %>
```

Comment:
I think you've uncovered the primary issue with the previous attempt, which I suspect is that we did not listen for `DOMContentLoaded` / `turbo:render` as pointed out by Alex. I think we might get away with using the deferred module approach, listening to `DOMContentLoaded` / `turbo:render`, at which point we can reuse the `window.OpenProject.theme` utility

I do share @myabc's concerns https://github.com/opf/openproject/pull/20709#pullrequestreview-3362710743 - particularly duplicating logic, we're better placed to avoid that at all costs.

> Another naive question: wouldn't it possible to initialise data attributes (`data-color-mode`) on `body` server-side?

Color modes are synced on the client-side based on the OS preferences, syncing on the server-side is not an option AFAICT.

Reply:
@bsatarnejad @myabc I just tried my own feedback and it didn't work 🤡 - we could also pair on this, 3 brains could work it out :)

Reply:
Yes agree, we can discuss it together.